### ISSUE TITLE
There should be a desktop app

### ISSUE THREAD

**Issue by yehosef** [Sefaria/Sefaria-Project#294]
Would be great if there was a "desktop" app using electron/chrome apps for offline for laptop use.  It could download all the books and ideally allow local saving of sheets/notes, etc until online for syncing.

**Comment by nsantacruz** [comment]
You can install our Android app on Windows using BlueStacks. It has an offline library, but no support for source sheets yet

**Comment by jeremieberrebi** [comment]
Android emulators are heavy... would be a lot better to have a Chrome App or a real software (I'm on Mac). Always better to learn offline

**Comment by nassan** [comment]
Okay I have confirmed that using https://github.com/jiahaog/nativefier the endpoint http://www.sefaria.org/texts can be packaged as an electron desktop app no problem.

In terms of offline functionality, the issue is that the data it would need to copy is in a mongodb behind a server. In order for this offline app to be effective it needs to just be installed and thats it. We cannot require an end user to install all of these dependencies (python, mongodb, etc.) that are required. Because of this we would need to copy the nosql database to a local serverless datastore, something like https://unqlite.org/.

Then, I think Electron uses nodejs to access the file system and Electron can also intercept urls, so theoretically the calls to the api can be intercepted and redirected to some code that attempts to get the resource from the umbedded nosql store.

**Comment by EliezerIsrael** [comment]
@nassan What if we made a read-only Mongo datastore available by url?  Would that work?

**Comment by nassan** [comment]
I don't think so. The client (for both web and electron) speak to api endpoints. Those endpoints are on the django server I believe? That is what it seemed like from reading through the source code. Although wasn't immediately obvious to me the path from request>router>read mongo>respond with json. Either way, this python code queries mongo for and returns back the correct json.

So having the mongo datastore available on a url doesn't necessarily help for 2 reasons:
1.  The client doesn't speak to mongo, it speaks to the django server
2. the url would have to be localhost for an offline app, and also require running mongo on the machine running the electron app.

Do you know the request>router>read mongo>respond with json works with the Sefaria backend? If I could understand that, I could possibly port it to Nodejs and read from an embedded version of the mongo datastore.

**Comment by yehosef** [comment]
Thanks for looking into this.  I've thought about this a little bit for my own project - here are some notes, references.  BTW, it's not practical to have an offline mongodb instance unless you want to bundle a VM - but it's far too heavy and clunky for the masses and won't deploy nicely as an app.

The most ideal approach is allow an offline approach which has the ability to make changes and sync, whatever could sync.  This is more work, but the best option for the users because it would increase the changes that people that are working offline can not just consume your content, but help create content (whether that is source sheets, translations, links, etc..).

For offline with sync - you can look at https://strongloop.com/strongblog/node-js-api-offline-sync-replication/ for some approaches.  Also couchdb supports this with pouchdb, but as that's not your backend, you'd have to consider if it's worth it (but take a look at [Mango](https://github.com/cloudant/mango).

If you want to have a read-only mongodb (it's not really mongo.. but looks like mongo) you can look into [NeDB](https://github.com/louischatriot/nedb) or [linvodb3](https://github.com/Ivshti/linvodb3).  The advantage of this is that you're already familiar with the queries need to get the data you want - the problem is that this code currently exists in python on the server so you'd have to port it to js and maintain it just for the desktop app - which is not good.  If the API was nodejs-driven already, you could reuse that code for an electron app - but that's not the case.  

I don't think it's a terrible idea to have a different storage mechanism between the a web and desktop client. I haven't looked into your client-side code, but you'll want to have the resource abstracted out so you're not referring to urls all over your code - just called some objects which will use an adapter for how to translate the object request into a data request - eg into a URL or localdb query, etc..  

Other links I found while writing this:
https://github.com/forbesmyester/SyncIt
http://stackoverflow.com/questions/22957902/best-practice-for-on-off-line-data-synchronization-using-angularjs-and-symfony-2
http://stackoverflow.com/questions/24384803/using-pouchdb-with-mongodb
https://github.com/mWater/minimongo

**Comment by nsantacruz** [comment]
Now that you're mentioning offline with sync, the way our iOS app works is that it downloads the library as a set of zip files. As you need to read a zip file it unzips it. The app is written in React Native which I assume cannot run as a desktop app, but all the basic structure of downloading our library as separate files and syncing is already built. 

We also have a SQL export of our MongoDB which we use on the Android app. I assume that can much more easily be built into a desktop app: https://github.com/Sefaria/Sefaria-SQL

**Comment by yehosef** [comment]
> Now that you're mentioning offline with sync

just so it's clear there are no surprises here.. I mentioned in the description

> It could download all the books and ideally allow local saving of sheets/notes, etc until online for syncing.

As I said, it's more difficult and might not be the first step.  But I think it has a lot of value, especially for this content.  

React native can work with destop apps, but it's not the main target:
https://github.com/Microsoft/react-native-windows
http://www.reactnative.com/build-os-x-desktop-apps-using-react-native/
But, you can also just build it using electron  - you don't need to use react native (just react)
([one example](https://medium.com/@Agro/developing-desktop-applications-with-electron-and-react-40d117d97564) - there are many)

If you have a process that works for the iOS or Android apps, it seems a natural place to start.  As I said, syncing add complications so you don't need to start there.  And it's possible, that you don't even ever really need to fully support it, but it would be great to have offline actions be pushed to the server when they come online (not exactly a sync - but has certain similarities..)

**Comment by nsantacruz** [comment]
Sorry @yehosef , you're right that syncing was clearly in the OP. I just realized that this problem has been mainly solved in our apps. 

If you do plan on pursuing this project, I would recommend going the route of converting the iOS React Native app into a desktop app. It already has syncing support in it and the zip files should work on all platforms out of the box. I'd be happy to help if need support understanding the iOS code to get this done.

**Comment by yehosef** [comment]
I'm not going to work on this for Sefaria.  I'm saying that I think there
is a big audience of sefaria users that would be interested in an
offline-able desktop app.  The idea of syncing is an extra bonus - but
there is still value without this.

But it's encouraging that it's doesn't seem complicated and should work out
of the box.  I hope sefaria is able to give it some attention.

**Comment by b3u** [comment]
I would be interested in taking a look at this project. I could do electron, but how does a PWA/native offline support sound?

**Comment by b3u** [comment]
@EliezerIsrael If you or someone could briefly outline this project, i would love to start on it.

**Comment by mikecg** [comment]
Count me in as someone who would really appreciate this. If someone wants to take the lead, I could help out.

**Comment by nsantacruz** [comment]
@b3u @mikecg I'll try to summarize this conversation this it's been going on a while. Sefaria doesn't see this project as a high enough priority to devote time to. That being said, we are happy if other developers want to take on this project and we are here to answer specific questions they have.

The easiest way to accomplish this project seems to be using React Native for Windows https://github.com/Microsoft/react-native-windows. The reason I say this is that we already have a mobile app written in React Native which has offline capabilities (https://github.com/Sefaria/Sefaria-iOS)

So the project would be to clone the Sefaria-iOS repo and modify it to work with React Native Windows.

**Comment by b3u** [comment]
@nsantacruz Thanks. I would rather go with Electron for Unix support, but that repo should be a great starting place. Would this project be under the sefaria org, or would I work on it under my personal acct?

**Comment by nsantacruz** [comment]
@b3u 

> I would rather go with Electron for Unix support

Fine with me. I know nothing about Electron so won't be able to give specific advice. I have no idea if React Native is compatible with Electron since Electron seems to be aimed at web apps. You may want to base your app off of this repo (https://github.com/Sefaria/Sefaria-Project) which is the code for our web app. That may lead to other issues since our web app depends on a running instance of MongoDB

> Would this project be under the sefaria org, or would I work on it under my personal acct

Your personal account for now

**Comment by b3u** [comment]
@mikecg Where do you propose we discuss this matter further?

**Comment by b3u** [comment]
@nsantacruz For the design - should I make it up as I go along, or make a proposal, etc.?

**Comment by nsantacruz** [comment]
@b3u This is your project from beginning to end. We're here to provide technical help regarding our software.

You can access all of our CSS styles from this repo (or simply by inspecting our site from your browser).

**Comment by b3u** [comment]
gotcha

**Comment by yehosef** [comment]
https://neutralino.js.org/ is a lighter-weight alternative to Electron  - just FYI.  

Great to see some interest/movement here.

**Comment by YonathanMeguira** [comment]
Hey I would like to re-open this thread, what about simply packaging with something like electron ?
I would really like taking this to desktop.  @nassan , @yehosef  @jeremieberrebi who's in ? where is it standing today ?

**Comment by YonathanMeguira** [comment]
can anyone give a rough estimation of how much data this represents (maybe a couple GB, lets not forget its all text)

**Comment by nsantacruz** [comment]
@YonathanMeguira it's about 3-4GB with just the texts and links. That number comes from adding up the Mongo collection sizes.
Sefaria already has an offline app for phones on both iOS and Android. You need internet to download the texts initially though.

**Comment by yehosef** [comment]
@nsantacruz - I'm assuming that's for all the texts and versions of the texts, which is probably not needed.  

If I'm not mistaken, the Sefaria app downloads the text into an SQLite database.  Instead, you could bundle that (especially a trimmed-down version) into the app.  You could also have a hybrid in which the app starts with a preloaded SQL database of the most needed and used texts and the user/app could download the rest if desired and internet-capable.  I would guestimate that you could get the 95% use case with less than 200MB of data.  (`find . -type f -name 'merged.json'  -exec du -ch {} +` in the `json` dir of the export repo gives around 460MB.)

I'm still primarily interested in a desktop app - I often want to learn in a Beis Midrash where I don't have internet access.  I don't mind downloading it when I'm online if that were an option.

**Comment by jeremieberrebi** [comment]
Same for me. Interested in a desktop app. I think it needs to be built
using https://www.electronjs.org/

The Android and iOS apps already have offline access

For the moment, I'm using the Android App with offline access on my Mac
using BlueStacks but it's really not ideal

- Jeremie Berrebi

ᐧ

**Comment by nsantacruz** [comment]
@yehosef the current Android and iOS apps use JSON files to store the text and links that come out to about 900MB. There's duplication there to make the app run at a reasonable speed. People are welcome to try to optimize the size further, but I feel that's not that important once you're talking about a desktop app where there should be ample space.

Reposting this to remind people what Sefaria thinks is the way to go here:

> @b3u @mikecg I'll try to summarize this conversation this it's been going on a while. Sefaria doesn't see this project as a high enough priority to devote time to. That being said, we are happy if other developers want to take on this project and we are here to answer specific questions they have.
> 
> The easiest way to accomplish this project seems to be using React Native for Windows https://github.com/Microsoft/react-native-windows. The reason I say this is that we already have a mobile app written in React Native which has offline capabilities (https://github.com/Sefaria/Sefaria-iOS)
> 
> So the project would be to clone the Sefaria-iOS repo and modify it to work with React Native Windows.

**Comment by YonathanMeguira** [comment]
@jeremieberrebi there is now a MAC version available.

Btw I saw what you wrote on chatgpt on mac. I got it on whatsapp.
In case you would like to try it.
058 628 8454

**Comment by jeremieberrebi** [comment]
Indeed. I'm using the iOS Edition on Mac already for a few months. Thanks

About ChatGPT on Whatsapp, I'm not using Whatsapp. Thanks :)

- Jeremie Berrebi

**Comment by JudahGabriel** [comment]
I work on the Microsoft Store, the app store for Windows, and we'd love to have Sefaria published as a Windows app.

If Electron is too much work, you could consider updating Sefaria to be a progressive web app (PWA). PWAs can be published in app stores like Google Play and Microsoft Store without the need for Electron packaging. 

Checking [pwabuilder.com](https://pwabuilder.com), I see Sefaria already has a [PWA manifest](https://www.sefaria.org/static/icons/manifest.json?v=xQzLlLp7oR). It just needs a few items added to the manifest:
![image](https://github.com/user-attachments/assets/7f0c8083-4fb1-4e92-8d54-d6ce48e3b35b). 

If those items could be added, PWABuilder can be used to publish to the app stores.

**Comment by saengel** [comment]
> I work on the Microsoft Store, the app store for Windows, and we'd love to have Sefaria published as a Windows app.
> 
> If Electron is too much work, you could consider updating Sefaria to be a progressive web app (PWA). PWAs can be published in app stores like Google Play and Microsoft Store without the need for Electron packaging.
> 
> [PWABuilder](https://pwabuilder.com) is Microsoft's tool that helps you build and publish PWAs in app stores. Running Sefaria's URL through PWABuilder, I see Sefaria already has a [PWA manifest](https://www.sefaria.org/static/icons/manifest.json?v=xQzLlLp7oR). It just needs a few items added to the manifest: ![image](https://private-user-images.githubusercontent.com/312936/364071675-7f0c8083-4fb1-4e92-8d54-d6ce48e3b35b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzODMzMzksIm5iZiI6MTcyNTM4MzAzOSwicGF0aCI6Ii8zMTI5MzYvMzY0MDcxNjc1LTdmMGM4MDgzLTRmYjEtNGU5Mi04ZDU0LWQ2Y2U0OGUzYjM1Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwM1QxNzAzNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMzg4Nzk2MmI0ZDMwMmYzYTRiYTU2YTVlMWVkNmNiNDhkNTdkYTg5ZGU5ZTZiMDJhYmY0ODkxMTA5YTE0OTc2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.pxwAWcXDtqmWTpU5INX5ylr1Eobo2K6SNdLVuD79pXU).
> 
> If those items could be added, PWABuilder can be used to publish to the app stores. If you'd like any help getting Sefaria published in the Microsoft Store, feel free to reach out to me juhimang at microsoft.

Thanks for letting us know!

**Comment by orxai** [comment]
Shana Tova friends :honey_pot: :apple: 
I'm working on this and made some progress.
Please check it out:
https://github.com/orxaicom/Sefaria-Desktop-Unofficial

**Comment by saengel** [comment]
Thanks for sending, we'll try to take a look!

**Comment by saengel** [comment]
@elico thank you for your feedback! We always appreciate hearing how we can best serve the needs of the community. Since we like to try and reserve this space for issues in the actual codebase, I'll close the issue - but feel free to follow up with the engineering team at developers@sefaria.org.

**Comment by saengel** [comment]
Another community-supported project powered by Sefaria's data also gives a desktop app experience, check out: https://github.com/kdroidFilter/Zayit
